plugins/optimization-detective/class-od-html-tag-processor.php

```diff
@@ -12,32 +12,563 @@
+	private $current_xpath = null;
+
+	/**
+	 * Whether the previous tag does not expect a closer.
+	 *
+	 * @since n.e.x.t
+	 * @var bool
+	 */
+	private $previous_tag_without_closer = false;
+
+	/**
+	 * Mapping of bookmark name to a list of HTML strings which will be inserted at the time get_updated_html() is called.
+	 *
+	 * @since n.e.x.t
+	 * @var array<string, string[]>
+	 */
+	private $buffered_text_replacements = array();
+
+	/**
+	 * Finds the next tag.
+	 *
+	 * Unlike the base class, this subclass disallows querying. This is to ensure the breadcrumbs can be tracked.
+	 * It will _always_ visit tag closers.
+	 *
+	 * @inheritDoc
+	 * @since n.e.x.t
+	 *
+	 * @param null $query Query.
+	 * @return bool Whether a tag was matched.
+	 *
+	 * @throws InvalidArgumentException If attempting to pass a query.
+	 */
+	public function next_tag( $query = null ): bool {
+		if ( null !== $query ) {
+			throw new InvalidArgumentException( esc_html__( 'Processor subclass does not support queries.', 'optimization-detective' ) );
+		}
+		return parent::next_tag( array( 'tag_closers' => 'visit' ) );
+	}
+
+	/**
+	 * Finds the next open tag.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool Whether a tag was matched.
+	 */
+	public function next_open_tag(): bool {
+		while ( $this->next_tag() ) {
+			if ( ! $this->is_tag_closer() ) {
+				return true;
+			}
 		}
+		return false;
+	}
 
-		$start = $this->bookmarks[ $bookmark ]->start;
+	/**
+	 * Whether the tag expects a closing tag.
+	 *
+	 * @see WP_HTML_Processor::expects_closer()
+	 * @since n.e.x.t
+	 *
+	 * @param string|null $tag_name Tag name, if not provided then the current tag is used. Optional.
+	 * @return bool Whether to expect a closer for the tag.
+	 */
+	public function expects_closer( ?string $tag_name = null ): bool {
+		if ( is_null( $tag_name ) ) {
+			$tag_name = $this->get_tag();
+		}
+		if ( is_null( $tag_name ) ) {
+			return false;
+		}
 
-		$this->lexical_updates[] = new WP_HTML_Text_Replacement(
-			$start,
-			0,
-			$html
+		return ! (
+			WP_HTML_Processor::is_void( $tag_name )
+			||
+			in_array( $tag_name, self::RAW_TEXT_TAGS, true )
 		);
+	}
+
+	/**
+	 * Finds the next token in the HTML document.
+	 *
+	 * @inheritDoc
+	 * @since n.e.x.t
+	 *
+	 * @return bool Whether a token was parsed.
+	 */
+	public function next_token(): bool {
+		$this->current_xpath = null; // Clear cache.
+		if ( ! parent::next_token() ) {
+			$this->open_stack_tags    = array();
+			$this->open_stack_indices = array();
+			return false;
+		}
+
+		$tag_name = $this->get_tag();
+		if ( null === $tag_name || $this->get_token_type() !== '#tag' ) {
+			return true;
+		}
+
+		if ( $this->previous_tag_without_closer ) {
+			array_pop( $this->open_stack_tags );
+		}
+
+		if ( ! $this->is_tag_closer() ) {
+
+			// Close an open P tag when a P-closing tag is encountered.
+			// TODO: There are quite a few more cases of optional closing tags: https://html.spec.whatwg.org/multipage/syntax.html#optional-tags
```

Comment:
This will be resolved when this class makes use of `WP_HTML_Processor` instead, as all of this will be handled automatically. But this will be WordPress 6.7.